### PR TITLE
OpenSSL 3 integration to avoid deprecated function calls.

### PR DESCRIPTION
**Pull request by pmesnier**
still working on DH and EC secret sharing functions.

### CODE REVIEW

File: dds/DCPS/security/SSL/DiffieHellman.cpp [hunk 1]
```diff
@@ -435,6 +698,7 @@ class ecdh_shared_secret_from_octets
 int ECDH_PRIME_256_V1_CEUM::compute_shared_secret(const DDS::OctetSeq& pub_key)
 {
   ecdh_shared_secret_from_octets secret(k_);
+
```
Comment (ClaytonCalabrese) [inline dds/DCPS/security/SSL/DiffieHellman.cpp:701]:
```suggestion
```

**Review by ClaytonCalabrese — CHANGES_REQUESTED** [review]

File: dds/DCPS/security/SSL/Certificate.cpp [hunk 2]
```diff
@@ -354,13 +354,21 @@ const char* Certificate::keypair_algo() const
...
+#ifndef OPENSSL_V_3_0
   cache_dsign_algo_impl() : pkey_(NULL), rsa_(NULL), ec_(NULL) {}
   ~cache_dsign_algo_impl()
   {
     EVP_PKEY_free(pkey_);
     RSA_free(rsa_);
     EC_KEY_free(ec_);
   }
+#else
+  cache_dsign_algo_impl() : pkey_(NULL) {}
```
Comment (iguessthislldo) [inline dds/DCPS/security/SSL/Certificate.cpp:366]:
```suggestion
  cache_dsign_algo_impl() : pkey_(0) {}
```

File: dds/DCPS/security/SSL/DiffieHellman.cpp [hunk 3]
```diff
@@ -195,22 +241,94 @@ class dh_shared_secret
...
     dst.length(len);
-
     len = DH_compute_key(dst.get_buffer(), pubkey, keypair);
     if (len < 0) {
       OPENDDS_SSL_LOG_ERR("DH_compute_key failed");
       dst.length(0u);
       return 1;
     }
+#else
+    if (!EVP_PKEY_get_utf8_string_param(keypair,"group", grp, glen, &glen)) {
```
Comment (iguessthislldo) [inline dds/DCPS/security/SSL/DiffieHellman.cpp:254]:
```suggestion
    if (!EVP_PKEY_get_utf8_string_param(keypair, "group", grp, glen, &glen)) {
```

File: dds/DCPS/security/SSL/DiffieHellman.cpp [hunk 4]
```diff
@@ -352,19 +482,57 @@ class ecdh_pubkey_as_octets
...
       OPENDDS_SSL_LOG_ERR("EC_POINT_point2oct failed");
       return 1;
     }
+#else
+    if (EVP_PKEY_todata(keypair, EVP_PKEY_KEYPAIR, &params) <= 0) {
+      OPENDDS_SSL_LOG_ERR("pkey to data failed");
+      return 1;
+    } else {
+      const char *gname = 0;
+      const unsigned char *pubbuf = 0;
```
Comment (iguessthislldo) [inline dds/DCPS/security/SSL/DiffieHellman.cpp:497]:
```suggestion
      const char* gname = 0;
      const unsigned char* pubbuf = 0;
```

File: dds/DCPS/security/SSL/DiffieHellman.cpp [hunk 5]
```diff
@@ -421,12 +606,92 @@ class ecdh_shared_secret_from_octets
...
 private:
+#ifndef OPENSSL_V_3_0
   EC_Handle keypair;
+#else
+  EVP_PKEY* keypair;
+  EVP_PKEY_CTX *ec_ctx;
+  EVP_PKEY_CTX *fd_ctx;
+  EVP_PKEY *peer;
+  OSSL_PARAM_BLD *param_bld;
+  OSSL_PARAM *params;
```
Comment (iguessthislldo) [inline dds/DCPS/security/SSL/DiffieHellman.cpp:693]:
```suggestion
  EVP_PKEY_CTX* ec_ctx;
  EVP_PKEY_CTX* fd_ctx;
  EVP_PKEY* peer;
  OSSL_PARAM_BLD* param_bld;
  OSSL_PARAM* params;
```

**Review by iguessthislldo — CHANGES_REQUESTED** [review]

**Comment by iguessthislldo** [comment]
Can you update this line to say that version 3 is also supported? https://github.com/objectcomputing/OpenDDS/blame/master/docs/dependencies.md#L187

File: dds/DCPS/security/SSL/DiffieHellman.cpp [hunk 6]
```diff
@@ -421,12 +598,92 @@ class ecdh_shared_secret_from_octets
...
     }
+#else
+    const char* grp = 0;
+    if (EVP_PKEY_todata(keypair, EVP_PKEY_PUBLIC_KEY, &params) <= 0) {
+      OPENDDS_SSL_LOG_ERR("pkey to data failed");
+      return 1;
+    } else {
+      for (OSSL_PARAM* p = params; grp == 0 && p != 0 && p->key != 0; p++) {
+        if (strcmp(p->key, "group") == 0) {
+          grp = (char *)p->data;
```
Comment (ClaytonCalabrese) [inline dds/DCPS/security/SSL/DiffieHellman.cpp:609]:
```suggestion
          grp = (char*)p->data;
```

**Review by ClaytonCalabrese — CHANGES_REQUESTED** [review]

**Review by iguessthislldo — APPROVED** [review]

**Review by ClaytonCalabrese — APPROVED** [review]

File: docs/dependencies.md [hunk 7]
```diff
@@ -184,7 +184,7 @@ security XML configuration files.
 ### OpenSSL
 
 OpenSSL is used for DDS Security for verifying security configurations and
-encryption and decryption. Version 1.0 and version 1.1 are supported.
+encryption and decryption. Versions 1.0, 1.1 and 3.0 are supported.
```
Comment (mitza-oci) [inline docs/dependencies.md:187]:
```suggestion
encryption and decryption. Versions 1.0, 1.1 and 3.0 (3.0.1 or later) are supported.
```

**Pull request merged**